mod-di-converter-storage-server/src/main/java/org/folio/services/importprofile/ProfileImportServiceImpl.java

```diff
@@ -0,0 +1,167 @@
+    if (profileSnapshot.getContentType() != JOB_PROFILE) {
+      String errorMessage = String.format(PROFILE_SNAPSHOT_INVALID_TYPE, profileSnapshot.getContentType(), JOB_PROFILE);
+      LOGGER.warn("importProfile:: {}", errorMessage);
+      return Future.failedFuture(new BadRequestException(errorMessage));
+    }
+    LOGGER.info("importProfile:: Started import for Job Profile with id {}", profileSnapshot.getProfileId());
+
+    try {
+      convertProfileSnapshotWrapperContent(profileSnapshot);
+    } catch (Exception e) {
+      return Future.failedFuture(new BadRequestException(String.format("Cannot map profile content, error: %s", e.getMessage())));
+    }
+
+    var profileTypeToSnapshots = getProfileTypeToSnapshot(profileSnapshot);
+    var profileTypesInSaveOrder = profileTypeToSnapshots.keySet().stream()
+      .sorted(Comparator.comparingInt(IMPORT_PROFILES_ORDER::indexOf))
+      .toList();
+
+    Future<List<Object>> saveFuture = Future.succeededFuture();
+    for (ProfileType profileType : profileTypesInSaveOrder) {
+      saveFuture = saveFuture.compose(v -> processProfilesSaving(profileTypeToSnapshots.get(profileType), profileType, okapiParams));
+    }
+    return saveFuture
+      .compose(v -> profileSnapshotService.constructSnapshot(profileSnapshot.getProfileId(), JOB_PROFILE, profileSnapshot.getProfileId(), okapiParams.getTenantId()));
+  }
+
+  private Future<List<Object>> processProfilesSaving(List<ProfileSnapshotWrapper> profileSnapshots, ProfileType profileType, OkapiConnectionParams okapiParams) {
+    List<Future<Object>> futures = new ArrayList<>();
+    var saveProfileFunction = profileTypeToSaveFunction.get(profileType);
+
+    profileSnapshots.forEach(profileSnapshot -> futures.add(saveProfileFunction.apply(profileSnapshot, okapiParams)));
+
+    return GenericCompositeFuture.all(futures).map(CompositeFuture::list);
+  }
+
+  private EnumMap<ProfileType, List<ProfileSnapshotWrapper>>  getProfileTypeToSnapshot(ProfileSnapshotWrapper snapshot) {
+    EnumMap<ProfileType, List<ProfileSnapshotWrapper>> profileTypeToSnapshots = new EnumMap<>(ProfileType.class);
+    addSnapshotsToMap(snapshot, profileTypeToSnapshots);
+    return profileTypeToSnapshots;
+  }
+
+  private void addSnapshotsToMap(ProfileSnapshotWrapper snapshot, EnumMap<ProfileType, List<ProfileSnapshotWrapper>> profileTypeToSnapshots) {
+    List<ProfileSnapshotWrapper> snapshots = profileTypeToSnapshots.computeIfAbsent(snapshot.getContentType(), k -> new ArrayList<>());
+
+    if (snapshots.stream().noneMatch(s -> s.getId().equals(snapshot.getId()))) {
+      snapshots.add(snapshot);
+    }
+
+    if (snapshot.getChildSnapshotWrappers() != null) {
+      for (ProfileSnapshotWrapper child : snapshot.getChildSnapshotWrappers()) {
+        addSnapshotsToMap(child, profileTypeToSnapshots);
+      }
+    }
+  }
+
+  private List<ProfileAssociation> formAddedRelations(ProfileSnapshotWrapper rootSnapshot, ProfileType rootContentType) {
```

Comment:
Please review this variant of the FormAddedRelations method.

```suggestion
private List<ProfileAssociation> formAddedRelations(ProfileSnapshotWrapper rootSnapshot, ProfileType rootContentType) {
    if (rootSnapshot.getChildSnapshotWrappers() == null) {
        return Collections.emptyList();
    }

    return rootSnapshot.getChildSnapshotWrappers().stream()
        .filter(childSnapshot -> shouldIncludeAssociation(rootContentType, childSnapshot))
        .flatMap(childSnapshot -> {
            List<ProfileAssociation> associations = new ArrayList<>();
            associations.add(createAssociation(rootSnapshot, childSnapshot));
            associations.addAll(formAddedRelations(childSnapshot, rootContentType));
            return associations.stream();
        })
        .collect(Collectors.toList());
}

private boolean shouldIncludeAssociation(ProfileType rootContentType, ProfileSnapshotWrapper childSnapshot) {
    return rootContentType == ACTION_PROFILE || childSnapshot.getContentType() != MAPPING_PROFILE;
}
```

Reply:
I think it is better to leave it as it is, thanks for your suggestion